core/store/src/trie/mod.rs

```diff
@@ -1711,17 +1712,51 @@ impl Trie {
         &self,
         boundary_account: &AccountId,
         retain_mode: RetainMode,
-        opts: AccessOptions,
-    ) -> Result<StateRoot, StorageError> {
-        let mut trie_update = TrieStorageUpdate::new(&self);
-        let root_node = self.move_node_to_mutable(&mut trie_update, &self.root, opts)?;
-        trie_update.retain_split_shard(boundary_account, retain_mode, opts);
-        #[cfg(test)]
-        {
-            self.memory_usage_verify(&trie_update, GenericNodeOrIndex::Updated(root_node.0));
+    ) -> Result<TrieChanges, StorageError> {
+        if self.memtries.is_some() {
+            // Get child trie_changes for all child memtries
+            let children_memtrie_changes = self
+                .children_memtries
+                .iter()
+                .map(|(shard_uid, memtrie)| {
+                    let inner_guard = memtrie.read().unwrap();
+                    let mut trie_update =
+                        inner_guard.update(self.root, TrackingMode::None).unwrap();
+                    trie_update.retain_split_shard(
+                        boundary_account,
+                        retain_mode,
+                        AccessOptions::DEFAULT,
+                    );
+                    (*shard_uid, trie_update.to_memtrie_changes_only())
+                })
+                .collect::<HashMap<_, _>>();
```

Comment:
I feel uneasy about HashMaps. If possible can you use a vector or BTreeMap? 
mini nit: I think a for look would be more readable but that's only personal preference.

Reply:
This isn't serialized anywhere. Should be fine to have the hashmap. We have the same in TrieChanges -> children_memtrie_changes. I'll change this to a for loop.

Reply:
I'm not worried about serialization, I'm worried that it may cause some processing to happen in non-deterministic order since hash map is randomized. It's unlikely that anything is broken right now but it's good practice to avoid hashmaps where possible IMO.